harvester/harvester
# Support vm snapshot

**Problem:**
Support in-cluster snapshot.

**Solution:**
Leverage [LH CSI Snapshot](https://github.com/longhorn/longhorn/issues/2534) to make it.

**Related Issue:**
https://github.com/harvester/harvester/issues/553

**Test plan:**
Setup:
* Create a harvester cluster from master-head.
* Create a VM `source-vm`.
* Run command `echo "123" > test.txt && sync` in `source-vm`.
* Configure Backup Target.
* Create a VMBackup `vm-backup-old`.

Case 1: New fields don't break old VMBackup
* Clone this branch and run following commands under `harvester/harvester/deploy/charts`:
```
> helm upgrade harvester-crd harvester-crd --namespace harvester-system
> helm upgrade harvester harvester --namespace harvester-system --set-string containers.apiserver.image.repository=frankyang/harvester --set-string containers.apiserver.image.tag=add-volumesnapshotnameclass-to-vmbackup-head --set-string webhook.image.repository=frankyang/harvester-webhook --set-string webhook.image.tag=add-volumesnapshotnameclass-to-vmbackup-head
```
* After upgrade done, check `vm-backup-old.spec.type` is `backup`, `vm-backup-old.status.csiDriverVolumeSnapshotClassNameMap` is not empty, and each `vm-backup-old.status.volumeBackup[]` has `csiDriverName`.
* Restore `vm-backup-old` to a new VM and replace `source-vm`. Both of VMs should have `test.txt` file with content `123`.
* Cleanup VMs and VMBackup.
* Set Backup Target as default.

Case 2: Create a snapshot for `source-vm`.
* Create a VM `source-vm`.
* Run command `echo "123" > test.txt && sync` in `source-vm`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineBackup
metadata:
  name: vm-snapshot
  namespace: default
spec:
  source:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: source-vm
  type: "snapshot"
```
* VMBackup `vm-snapshot` should be created successfully.

Case 3: Restore `vm-snapshot` to a new VM `vm-from-snapshot`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineRestore
metadata:
  name: restore-vm-from-snapshot
  namespace: default
spec:
  newVM: true
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: vm-from-snapshot
  virtualMachineBackupName: vm-snapshot
  virtualMachineBackupNamespace: default
```
* A new VM `vm-from-snapshot` should be created successfully.
* Run command `cat test.txt` in `vm-from-snapshot` and it should show content `123`.

Case 4: Replace `source-vm` with `vm-snapshot` which `deletePolicy` is not `retain`. (Negative case)
* Stop a VM `source-vm`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineRestore
metadata:
  name: replace-vm-from-snapshot
  namespace: default
spec:
  newVM: false
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: source-vm
  virtualMachineBackupName: vm-snapshot
  virtualMachineBackupNamespace: default
```
* We should get an error message like `Delete policy with backup type snapshot for replacing VM is not supported`.

Case 5: Replace `source-vm` with `vm-snapshot` which `deletePolicy` is `retain`.
* Stop a VM `source-vm`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineRestore
metadata:
  name: replace-vm-from-snapshot
  namespace: default
spec:
  newVM: false
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: source-vm
  virtualMachineBackupName: vm-snapshot
  virtualMachineBackupNamespace: default
  deletionPolicy: retain
```
* `source-vm` starts again and its volume name is `restore-vm-snapshot-...`.
* Run command `cat test.txt` in `source-vm` and it should show content `123`.

Case 6: Restore `vm-snapshot` to a new VM `vm-from-snapshot-2` when the source PVC is detached.
* Manually detach `source-vm-disk-0-...` PVC in LH dashboard.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineRestore
metadata:
  name: restore-vm-from-snapshot-2
  namespace: default
spec:
  newVM: true
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: vm-from-snapshot-2
  virtualMachineBackupName: vm-snapshot
  virtualMachineBackupNamespace: default
```
* A new VM `vm-from-snapshot-2` should be created successfully.
* Run command `cat test.txt` in `vm-from-snapshot-2` and it should show content `123`.

Case 7: Detached PVC can be taken a snapshot.
* Stop `source-vm`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineBackup
metadata:
  name: vm-snapshot-2
  namespace: default
spec:
  source:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: source-vm
  type: "snapshot"
```
* A new `vm-snapshot-2` VMBackup is created.

Case 8: Delete VM and related VMBackup will be deleted (no matter whether with disk or not).
* Delete `source-vm` with disks.
* `vm-snapshot` and `vm-snapshot-2` VMBackup are deleted too.
* `vm-snapshot-volume-source-vm-disk0-xxx` VolumeSnapshot is still existent because related PVC `source-vm-disk0-xxx` is not deleted. Also the VolumeSnapshot's owner reference is changed to `source-vm-disk0-xxx` PVC.

Case 9: Delete PVC and related VolumeSnapshot will be deleted.
* Delete `source-vm-disk-0-xxx` PVC.
* `vm-snapshot-volume-source-vm-disk-0-xxx` VolumeSnapshot is deleted too.

Case 10: Create a backup for `new-vm`.
* Create a `new-vm` again and run the command `echo "123" > test.txt && sync` in it.
* Setup backup target.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineBackup
metadata:
  name: vm-backup
  namespace: default
spec:
  source:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: new-vm
  type: "backup"
```
* VMBackup `vm-backup` should be created successfully.

Case 11: Restore `vm-backup` to a new VM `vm-from-backup`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineRestore
metadata:
  name: restore-vm-from-backup
  namespace: default
spec:
  newVM: true
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: vm-from-backup
  virtualMachineBackupName: vm-backup
  virtualMachineBackupNamespace: default
```
* A new VM `vm-from-backup` should be created successfully.
* Run command `cat test.txt` in `vm-from-snapshot` and it should show content `123`.

Case 12: Replace `new-vm` with `vm-backup`.
* Run command `echo "456" > test.txt && sync` in `new-vm`.
* Stop `new-vm`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineRestore
metadata:
  name: replace-vm-from-backup
  namespace: default
spec:
  newVM: false
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: new-vm
  virtualMachineBackupName: vm-backup
  virtualMachineBackupNamespace: default
```
* VM `new-vm` should be restored successfully.
* Run command `cat test.txt` in `new-vm` and it should show content `123`.

**Additional context**
Currently, LH CSI snapshot has a limitation - we need to attach volume by ourselves. https://github.com/longhorn/longhorn/issues/3726 merge.
